ignite/handlers/state_param_scheduler.py

```diff
@@ -168,45 +165,71 @@ class LinearStateScheduler(StateParamScheduler):
+            pwlinear_state_parameter_scheduler.attach(engine, Events.EPOCH_COMPLETED)
+
+            # basic handler to print scheduled state parameter
+            engine.add_event_handler(Events.EPOCH_COMPLETED, lambda _ : print(engine.state.pw_linear_scheduled_param))
+
+            engine.run([0] * 8, max_epochs=40)
+            #
+            # Sets the state parameter value to 0.5 over the first 10 iterations, then decreases linearly from 0.5 to
+            # 0.45 between 10th and 20th iterations. Next there is a jump to 0.3 at the 21st iteration and the state
+            # parameter value decreases linearly from 0.3 to 0.1 between 21st and 30th iterations and remains 0.1 until
+            # the end of the iterations.
+            #
 
     .. versionadded:: 0.6.0
     """
 
     def __init__(
-        self,
-        initial_value: float,
-        step_constant: int,
-        step_max_value: int,
-        max_value: float,
-        param_name: str,
-        save_history: bool = False,
+        self, milestones_values: List[Tuple[int, float]], param_name: str, save_history: bool = False,
     ):
-        super(LinearStateScheduler, self).__init__(param_name, save_history)
-        self.initial_value = initial_value
-        self.step_constant = step_constant
-        self.step_max_value = step_max_value
-        self.max_value = max_value
-        self._state_attrs += ["initial_value", "step_constant", "step_max_value", "max_value"]
+        super(PwLinearStateScheduler, self).__init__(param_name, save_history)
+
+        values = []  # type: List[float]
+        milestones = []  # type: List[int]
+        for pair in milestones_values:
+            if not isinstance(pair, tuple) or len(pair) != 2:
+                raise ValueError("Argument milestones_values should be a list of pairs (milestone, param_value)")
+            if not isinstance(pair[0], numbers.Integral):
+                raise TypeError(f"Value of a milestone should be integer, but given {type(pair[0])}")
+            if len(milestones) > 0 and pair[0] < milestones[-1]:
+                raise ValueError(
+                    f"Milestones should be increasing integers, but given {pair[0]} is smaller "
+                    f"than the previous milestone {milestones[-1]}"
+                )
+            milestones.append(pair[0])
+            values.append(pair[1])
 
-    def get_param(self) -> Union[List[float], float]:
-        if self.event_index <= self.step_constant:
-            delta = 0.0
-        elif self.event_index > self.step_max_value:
-            delta = self.max_value - self.initial_value
-        else:
-            delta = (
-                (self.max_value - self.initial_value)
-                / (self.step_max_value - self.step_constant)
-                * (self.event_index - self.step_constant)
+        self.values = values
+        self.milestones = milestones
+        self._index = 0
+        self._state_attrs += ["initial_value", "milestones", "_index"]
```

Comment:
```suggestion
        self._state_attrs += ["values", "milestones", "_index"]
```

Reply:
yes of course, missed that ...

Reply:
Do you check save/load state dict for these classes ? Seems like not, but we have to do that

Reply:
Yes you're right will add this to the tests